### PR TITLE
Upgrade to latest remark version

### PR DESCRIPTION
**Pull request by j-f1**
## Description

Updgrades to latest Remark and tweaks our custom parsing things to work with it.

## TODO

- [ ] MDX
- [ ] evaluate snapshot changes and determine if they’re bugs or improvements
- [ ] fix `TypeError: Cannot read properties of undefined (reading '1') at subcontent (/micromark-util-subtokenize/index.js:165:35)`

## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Have you seen https://github.com/prettier/prettier/issues/12209#issuecomment-1266950613?

**Comment by j-f1** [comment]
I definitely saw it when it was posted but forgot about it. Closing since they're working on it and will probably do a better job than me :)

**Comment by remcohaszing** [comment]
See #12209

I would love to help out if you decide to continue this PR and have any questions. My first suggestion is: add [`remark-gfm`](https://github.com/remarkjs/remark-gfm)

File: src/language-markdown/parser-markdown.js [hunk 1]
```diff
@@ -26,20 +27,26 @@ import looseItems from "./unified-plugins/loose-items.js";
...
+    .use(liquid)
+    // .use(isMDX ? htmlToJsx : identity)
+    .use(wikiLink);
+  // .use(looseItems);
+
+  return async (text) => {
+    const { frontMatter, content } = parseFrontMatter(text);
+    if (frontMatter) {
+      const spacer = frontMatter.raw.replace(/./g, " ");
+      const ast = await processor.run(processor.parse(spacer + content));
```
[the remark this answers is not in the document]
Comment (j-f1) [reply src/language-markdown/parser-markdown.js:45]: Oops! Will investigate

**Comment by tats-u** [comment]
I tried to rebase this branch to the latest `main` because it's based on 3.0.0-beta.4, which is too old, but bumped into so many numbers of conflicts that I gave up.

**Comment by fisker** [comment]
@tats-u  Merged main into this branch, want continue work on it?

**Comment by tats-u** [comment]
@fisker I'm glad this PR turned to be sustainable now.
Unfortunately I don't know which version of MDX will be available after this PR, what problems this PR is stuck with, or what percent is the progress of this PR.
Possibly I might be able to contribute by a simple task...?

**Comment by fisker** [comment]
You can push clone and push this to your fork. 

My plan is keep MDX unchanged, upgrade markdown first. Then we'll upgrade mdx.

**Comment by fisker** [comment]
> You took the action merge, which makes it difficult to rebase next time, instead of rebase.

Because it's hard to rebase, we can also do merge next time.

**Comment by fisker** [comment]
The old mdx parse logic is kept in `src/language-markdown/parser-mdx.js`, we can leave it alone for now.
The hard part is we need change markdown print part. If it's necessary, we can have different print logic for mdx (keep the old unchanged) and markdown by checking `options.parser`.

It has been a while I can't remember the exact problem, but if I remember correctly the problem I run into is that I can't the real value of indented `code` and `blockquote` .

**Comment by seiyab** [comment]
Mentioning https://github.com/prettier/prettier/issues/11828 just to get linked.

**Comment by fisker** [comment]
@j-f1 I think I have made great progress here now.

In this PR, let's focus on the markdown parsing, I kept the `mdx` parser untouched, it still using `unified@9` and `remark-parse@8`.

I've created `remark-11` branch, let's merge this one into it, and then start working on the `mdx` parser, then we merge together into `main` branch.

 @seiyab Thanks for the liquid plugin from #16134

We still need go through the test changes to make sure they are right. 

I'm not going to continue work on this recently. Welcome to continue the work or review if interested. @j-f1 @seiyab @remcohaszing @tats-u

**Comment by seiyab** [comment]
Thank you for your work.
I'm wondering how can we manage that which format changes are valid and which ones are invalid well...

**Comment by fisker** [comment]
The AST_COMPARE test can test if the AST changed. Eg: https://github.com/prettier/prettier/actions/runs/9626611049/job/26552757384?pr=13676#step:6:2509

**Comment by seiyab** [comment]
Sounds good! But maybe some wrong parsing can be missed? (technically unchanged wrong AST can exist)

**Comment by remcohaszing** [comment]
I would put some faith in the parser that it produces correct ASTs.

**Comment by fisker** [comment]
> submitting PR against prettier:remark-11 directly fine?

You can do that

**Comment by seiyab** [comment]
I'm sorry but I'm not working on this now.
Feel free to tackle this if someone is interested in it.

**Comment by fisker** [comment]
Continue in #18277

**Pull request closed without merging**